### PR TITLE
Make executable document directives an Option

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Resolves #12 

This transforms the `executable` document directives into an `Option<Directives>`, currently this already reduces the overall memory footprint of the AST by returning `None` when there are no directives present on the ExecutableDefinition. The allocations however do happen when we are parsing, a future optimization here would be to peek and see whether we actually have directives, if there are none we can skip the allocations completely. I saw that we have the same redundant `vec` allocation in the `SchemaDefinition`, this does mean that we have a slight discrepancy when there's `None` in the `definition.directives` between `executable` and `schema` because we add an additional bail case for an empty `vec`.

Do note that this is in theory a breaking change, this would make for `bluejay-core` 0.2.0

### CODE REVIEW

File: bluejay-parser/src/ast/executable/field.rs [hunk 1]
```diff
@@ -46,7 +48,11 @@ impl<'a> FromTokens<'a> for Field<'a> {
             alias,
             name,
             arguments,
-            directives,
+            directives: if directives.len() > 0 {
```
Comment (JoviDeCroock) [inline bluejay-parser/src/ast/executable/field.rs:51]: For instance here we could on line 34 peek and see if we get a directive or not, if we don't we can just return None

Comment (JoviDeCroock) [reply bluejay-parser/src/ast/executable/field.rs:51]: `try_from_tokens` does however not allow us to do this optimization on the AST and kind of seems to hide it behind an Option instead. I started with that approach but it felt a bit counter-intuitive

File: bluejay-printer/src/executable/fragment_spread.rs [hunk 2]
```diff
@@ -24,11 +24,15 @@ impl<'a, T: FragmentSpread> Display for FragmentSpreadPrinter<'a, T> {
...
+        if let Some(directives) = fragment_spread.directives() {
+            write!(
+                f,
+                "...{}{}",
+                fragment_spread.name(),
+                DirectivesPrinter::new(directives)
+            )
+        } else {
+            write!(f, "...{}", fragment_spread.name(),)
+        }
```
[the remark this answers is not in the document]
Comment (JoviDeCroock) [reply bluejay-printer/src/executable/fragment_spread.rs]: My thinking was that that was less performant so I avoided it, will add

File: bluejay-parser/src/ast/executable/field.rs [hunk 3]
```diff
@@ -29,12 +29,13 @@ impl<'a> FromTokens<'a> for Field<'a> {
...
         let arguments = VariableArguments::try_from_tokens(tokens).transpose()?;
-        let directives = VariableDirectives::from_tokens(tokens)?;
+        let directives = VariableDirectives::try_from_tokens(tokens).transpose()?;
         let selection_set = SelectionSet::try_from_tokens(tokens).transpose()?;
         let start_span = alias.as_ref().unwrap_or(&name).span();
+        let directives_span = directives.as_ref().map(|x| x.span());
         let end_span = if let Some(selection_set) = &selection_set {
             selection_set.span()
-        } else if let Some(directive_span) = directives.span() {
+        } else if let Some(Some(directive_span)) = directives_span {
```
Comment (JoviDeCroock) [inline bluejay-parser/src/ast/executable/field.rs:38]: This is a bit uglier with `try_from_tokens` but I guess it's worth the sacrifice 😅

**Review by adampetro — APPROVED** [review]
This is great, thanks! Do you mind squashing the commits down, or using squash when merging?

**Pull request merged**